bug: react, modal does not wait for contents to mount before opening

### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

v7.x

### Current Behavior

When opening the modal, there's a high chance the animation lags and is not smooth. 

### Expected Behavior

When opening the modal, the animation should never lag. It should always be smooth.

### Steps to Reproduce

1. Clone the repo available at [https://github.com/galash13/ionic-react-modal-issue](https://github.com/galash13/ionic-react-modal-issue)
2. Open the Xcode project and run on a device or simulator
3. Try opening/closing the modal. If the issue doesn't immediately appear, try again more times and it will. 

The modal is being opened and closed with the `isOpen` method described in the [docs](https://ionicframework.com/docs/api/modal#using-isopen)

Below you can find a few screen recordings of the issue on different devices

https://user-images.githubusercontent.com/56839513/235364027-95dbe817-a8a5-4984-9710-9649db649647.mp4


https://user-images.githubusercontent.com/56839513/235364035-a158a0e1-9f5f-43be-a309-c3024014fd0d.mp4


https://user-images.githubusercontent.com/56839513/235364847-28cbcbff-1b86-4e93-8511-c8564595e475.mp4



### Code Reproduction URL

https://github.com/galash13/ionic-react-modal-issue

### Ionic Info

[WARN] Error loading @capacitor/android package.json: Error: Cannot find module '@capacitor/android/package.json'
       
       Require stack:
       - /Users/galash13/.nvm/versions/node/v19.9.0/lib/node_modules/@ionic/cli/lib/project/index.js
       - /Users/galash13/.nvm/versions/node/v19.9.0/lib/node_modules/@ionic/cli/lib/index.js
       - /Users/galash13/.nvm/versions/node/v19.9.0/lib/node_modules/@ionic/cli/index.js
       - /Users/galash13/.nvm/versions/node/v19.9.0/lib/node_modules/@ionic/cli/bin/ionic

Ionic:

   Ionic CLI       : 7.0.1 (/Users/galash13/.nvm/versions/node/v19.9.0/lib/node_modules/@ionic/cli)
   Ionic Framework : @ionic/react 7.0.4

Capacitor:

   Capacitor CLI      : 4.8.0
   @capacitor/android : not installed
   @capacitor/core    : 4.8.0
   @capacitor/ios     : 4.8.0

Utility:

   cordova-res : not installed globally
   native-run  : 1.7.2

System:

   NodeJS : v19.9.0 (/Users/galash13/.nvm/versions/node/v19.9.0/bin/node)
   npm    : 9.6.3
   OS     : macOS 13 Ventura

### Additional Information

_No response_

--- a/angular/src/directives/overlays/modal.ts
+++ b/angular/src/directives/overlays/modal.ts
@@ -116,7 +116,7 @@ export class IonModal {
   constructor(c: ChangeDetectorRef, r: ElementRef, protected z: NgZone) {
     this.el = r.nativeElement;
 
-    this.el.addEventListener('willPresent', () => {
+    this.el.addEventListener('ionMount', () => {
       this.isCmpOpen = true;
       c.detectChanges();
     });
@@ -124,7 +124,6 @@ export class IonModal {
       this.isCmpOpen = false;
       c.detectChanges();
     });
-
     proxyOutputs(this, this.el, [
       'ionModalDidPresent',
       'ionModalWillPresent',

--- a/core/src/components.d.ts
+++ b/core/src/components.d.ts
@@ -5831,6 +5831,10 @@ declare namespace LocalJSX {
          */
         "onIonModalWillPresent"?: (event: IonModalCustomEvent<void>) => void;
         /**
+          * Emitted before the modal has presented, but after the component has been mounted in the DOM. This event exists so iOS can run the entering transition properly
+         */
+        "onIonMount"?: (event: IonModalCustomEvent<void>) => void;
+        /**
           * Emitted before the modal has dismissed. Shorthand for ionModalWillDismiss.
          */
         "onWillDismiss"?: (event: IonModalCustomEvent<OverlayEventDetail>) => void;

--- a/core/src/components/modal/modal.tsx
+++ b/core/src/components/modal/modal.tsx
@@ -31,7 +31,7 @@ import {
 } from '../../utils/overlays';
 import type { OverlayEventDetail } from '../../utils/overlays-interface';
 import { getClassMap } from '../../utils/theme';
-import { deepReady } from '../../utils/transition';
+import { deepReady, waitForMount } from '../../utils/transition';
 
 import { iosEnterAnimation } from './animations/ios.enter';
 import { iosLeaveAnimation } from './animations/ios.leave';
@@ -316,6 +316,16 @@ export class Modal implements ComponentInterface, OverlayInterface {
    */
   @Event({ eventName: 'didDismiss' }) didDismissShorthand!: EventEmitter<OverlayEventDetail>;
 
+  /**
+   * Emitted before the modal has presented, but after the component
+   * has been mounted in the DOM.
+   * This event exists so iOS can run the entering
+   * transition properly
+   *
+   * @internal
+   */
+  @Event() ionMount!: EventEmitter<void>;
+
   breakpointsChanged(breakpoints: number[] | undefined) {
     if (breakpoints !== undefined) {
       this.sortedBreakpoints = breakpoints.sort((a, b) => a - b);
@@ -443,7 +453,29 @@ export class Modal implements ComponentInterface, OverlayInterface {
 
     const { inline, delegate } = this.getDelegate(true);
     this.usersElement = await attachComponent(delegate, el, this.component, ['ion-page'], this.componentProps, inline);
-    hasLazyBuild(el) && (await deepReady(this.usersElement));
+
+    /**
+     * When using the lazy loaded build of Stencil, we need to wait
+     * for every Stencil component instance to be ready before presenting
+     * otherwise there can be a flash of unstyled content. With the
+     * custom elements bundle we need to wait for the JS framework
+     * mount the inner contents of the overlay otherwise WebKit may
+     * get the transition incorrect.
+     */
+    if (hasLazyBuild(el)) {
+      await deepReady(this.usersElement);
+      /**
+       * If keepContentsMounted="true" then the
+       * JS Framework has already mounted the inner
+       * contents so there is no need to wait.
+       * Otherwise, we need to wait for the JS
+       * Framework to mount the inner contents
+       * of this component.
+       */
+    } else if (!this.keepContentsMounted) {
+      this.ionMount.emit();
+      await waitForMount();
+    }
 
     writeTask(() => this.el.classList.add('show-modal'));
 

--- a/core/src/components/modal/modal.tsx
+++ b/core/src/components/modal/modal.tsx
@@ -454,6 +454,8 @@ export class Modal implements ComponentInterface, OverlayInterface {
     const { inline, delegate } = this.getDelegate(true);
     this.usersElement = await attachComponent(delegate, el, this.component, ['ion-page'], this.componentProps, inline);
 
+    this.ionMount.emit();
+
     /**
      * When using the lazy loaded build of Stencil, we need to wait
      * for every Stencil component instance to be ready before presenting
@@ -473,7 +475,6 @@ export class Modal implements ComponentInterface, OverlayInterface {
        * of this component.
        */
     } else if (!this.keepContentsMounted) {
-      this.ionMount.emit();
       await waitForMount();
     }
 

--- a/core/src/components/popover/popover.tsx
+++ b/core/src/components/popover/popover.tsx
@@ -10,7 +10,7 @@ import { BACKDROP, dismiss, eventMethod, focusFirstDescendant, prepareOverlay, p
 import type { OverlayEventDetail } from '../../utils/overlays-interface';
 import { isPlatform } from '../../utils/platform';
 import { getClassMap } from '../../utils/theme';
-import { deepReady } from '../../utils/transition';
+import { deepReady, waitForMount } from '../../utils/transition';
 
 import { iosEnterAnimation } from './animations/ios.enter';
 import { iosLeaveAnimation } from './animations/ios.leave';
@@ -455,61 +455,57 @@ export class Popover implements ComponentInterface, PopoverInterface {
       this.componentProps,
       inline
     );
-    hasLazyBuild(el) && (await deepReady(this.usersElement));
 
     if (!this.keyboardEvents) {
       this.configureKeyboardInteraction();
     }
     this.configureDismissInteraction();
 
-    this.ionMount.emit();
-
-    return new Promise((resolve) => {
+    /**
+     * When using the lazy loaded build of Stencil, we need to wait
+     * for every Stencil component instance to be ready before presenting
+     * otherwise there can be a flash of unstyled content. With the
+     * custom elements bundle we need to wait for the JS framework
+     * mount the inner contents of the overlay otherwise WebKit may
+     * get the transition incorrect.
+     */
+    if (hasLazyBuild(el)) {
+      await deepReady(this.usersElement);
       /**
-       * Wait two request animation frame loops before presenting the popover.
-       * This allows the framework implementations enough time to mount
-       * the popover contents, so the bounding box is set when the popover
-       * transition starts.
-       *
-       * On Angular and React, a single raf is enough time, but for Vue
-       * we need to wait two rafs. As a result we are using two rafs for
-       * all frameworks to ensure the popover is presented correctly.
+       * If keepContentsMounted="true" then the
+       * JS Framework has already mounted the inner
+       * contents so there is no need to wait.
+       * Otherwise, we need to wait for the JS
+       * Framework to mount the inner contents
+       * of this component.
        */
-      raf(() => {
-        raf(async () => {
-          this.currentTransition = present<PopoverPresentOptions>(
-            this,
-            'popoverEnter',
-            iosEnterAnimation,
-            mdEnterAnimation,
-            {
-              event: event || this.event,
-              size: this.size,
-              trigger: this.triggerEl,
-              reference: this.reference,
-              side: this.side,
-              align: this.alignment,
-            }
-          );
+    } else if (!this.keepContentsMounted) {
+      this.ionMount.emit();
+      await waitForMount();
+    }
 
-          await this.currentTransition;
-
-          this.currentTransition = undefined;
-
-          /**
-           * If popover is nested and was
-           * presented using the "Right" arrow key,
-           * we need to move focus to the first
-           * descendant inside of the popover.
-           */
-          if (this.focusDescendantOnPresent) {
-            focusFirstDescendant(this.el, this.el);
-          }
-
-          resolve();
-        });
-      });
+    this.currentTransition = present<PopoverPresentOptions>(this, 'popoverEnter', iosEnterAnimation, mdEnterAnimation, {
+      event: event || this.event,
+      size: this.size,
+      trigger: this.triggerEl,
+      reference: this.reference,
+      side: this.side,
+      align: this.alignment,
     });
+
+    await this.currentTransition;
+
+    this.currentTransition = undefined;
+
+    /**
+     * If popover is nested and was
+     * presented using the "Right" arrow key,
+     * we need to move focus to the first
+     * descendant inside of the popover.
+     */
+    if (this.focusDescendantOnPresent) {
+      focusFirstDescendant(this.el, this.el);
+    }
   }
 
   /**

--- a/core/src/components/popover/popover.tsx
+++ b/core/src/components/popover/popover.tsx
@@ -461,6 +461,8 @@ export class Popover implements ComponentInterface, PopoverInterface {
     }
     this.configureDismissInteraction();
 
+    this.ionMount.emit();
+
     /**
      * When using the lazy loaded build of Stencil, we need to wait
      * for every Stencil component instance to be ready before presenting
@@ -480,7 +482,6 @@ export class Popover implements ComponentInterface, PopoverInterface {
        * of this component.
        */
     } else if (!this.keepContentsMounted) {
-      this.ionMount.emit();
       await waitForMount();
     }
 

--- a/core/src/utils/transition/index.ts
+++ b/core/src/utils/transition/index.ts
@@ -211,8 +211,8 @@ export const lifecycle = (el: HTMLElement | undefined, eventName: string) => {
  * we need to wait two rafs. As a result we are using two rafs for
  * all frameworks to ensure contents are mounted.
  */
-export const waitForMount = () => {
-  return new Promise((resolve) => raf(raf(resolve)));
+export const waitForMount = (): Promise<void> => {
+  return new Promise((resolve) => raf(() => raf(() => resolve())));
 };
 
 export const deepReady = async (el: any | undefined): Promise<void> => {

--- a/core/src/utils/transition/index.ts
+++ b/core/src/utils/transition/index.ts
@@ -198,6 +198,23 @@ export const lifecycle = (el: HTMLElement | undefined, eventName: string) => {
   }
 };
 
+/**
+ * Wait two request animation frame loops.
+ * This allows the framework implementations enough time to mount
+ * the user-defined contents. This is often needed when using inline
+ * modals and popovers that accept user components. For popover,
+ * the contents must be mounted for the popover to be sized correctly.
+ * For modals, the contents must be mounted for iOS to run the
+ * transition correctly.
+ *
+ * On Angular and React, a single raf is enough time, but for Vue
+ * we need to wait two rafs. As a result we are using two rafs for
+ * all frameworks to ensure contents are mounted.
+ */
+export const waitForMount = () => {
+  return new Promise((resolve) => raf(raf(resolve)));
+};
+
 export const deepReady = async (el: any | undefined): Promise<void> => {
   const element = el as any;
   if (element) {

